airbyte-integrations/connectors/destination-kafka/src/main/java/io/airbyte/integrations/destination/kafka/KafkaRecordConsumer.java

```diff
@@ -0,0 +1,156 @@
+  private static final Logger LOGGER = LoggerFactory.getLogger(KafkaRecordConsumer.class);
+
+  private final String topicPattern;
+  private final Map<AirbyteStreamNameNamespacePair, String> topicMap;
+  private final KafkaProducer<String, JsonNode> producer;
+  private final boolean sync;
+  private final boolean transactionalProducer;
+  private final ConfiguredAirbyteCatalog catalog;
+  private final Callback callback;
+  private final Consumer<AirbyteMessage> outputRecordCollector;
+  private final NamingConventionTransformer nameTransformer;
+
+  private AirbyteMessage lastStateMessage = null;
+
+  public KafkaRecordConsumer(KafkaDestinationConfig kafkaDestinationConfig,
+                             ConfiguredAirbyteCatalog catalog,
+                             Consumer<AirbyteMessage> outputRecordCollector,
+                             NamingConventionTransformer nameTransformer) {
+    this.topicPattern = kafkaDestinationConfig.getTopicPattern();
+    this.topicMap = new HashMap<>();
+    this.producer = kafkaDestinationConfig.getProducer();
+    this.sync = kafkaDestinationConfig.isSync();
+    this.transactionalProducer = kafkaDestinationConfig.isTransactionalProducer();
+    this.catalog = catalog;
+    this.callback = (metadata, exception) -> {
+      if (exception != null) {
+        // TODO improve error management
+        LOGGER.error("Error sending message to topic", exception);
+      }
+    };
+    this.outputRecordCollector = outputRecordCollector;
+    this.nameTransformer = nameTransformer;
+  }
+
+  @Override
+  protected void startTracked() {
+    Map<AirbyteStreamNameNamespacePair, String> mapped = catalog.getStreams().stream()
+        .map(stream -> AirbyteStreamNameNamespacePair.fromAirbyteSteam(stream.getStream()))
+        .collect(Collectors.toMap(Function.identity(),
+            pair -> nameTransformer.getIdentifier(topicPattern
+                .replaceAll("\\{namespace}", Optional.ofNullable(pair.getNamespace()).orElse(""))
+                .replaceAll("\\{stream}", Optional.ofNullable(pair.getName()).orElse("")))));
+    topicMap.putAll(mapped);
+
+    if (transactionalProducer) {
+      producer.initTransactions();
+    }
+  }
+
+  @Override
+  protected void acceptTracked(AirbyteMessage airbyteMessage) throws Exception {
+    if (airbyteMessage.getType() == AirbyteMessage.Type.STATE) {
+      lastStateMessage = airbyteMessage;
+    } else if (airbyteMessage.getType() == AirbyteMessage.Type.RECORD) {
+      final AirbyteRecordMessage recordMessage = airbyteMessage.getRecord();
+
+      final String topic = topicMap.get(AirbyteStreamNameNamespacePair.fromRecordMessage(recordMessage));
+      final String key = UUID.randomUUID().toString();
+      final JsonNode value = Jsons.jsonNode(ImmutableMap.of(
+          JavaBaseConstants.COLUMN_NAME_AB_ID, key,
+          JavaBaseConstants.COLUMN_NAME_EMITTED_AT, recordMessage.getEmittedAt(),
+          JavaBaseConstants.COLUMN_NAME_DATA, recordMessage.getData()));
```

Comment:
if a user is writing to a hardcoded stream, they have no way of knowing which stream this came from. Should we include the stream name in the output record?

Reply:
Good point! I just added it